More refactorings

dmke opened the pull request:
Hey,

whilst using this more, I found I needed a more conveniant interface.

The first part in this commit series will introduce a `sendmail.Option` type and a `sendmail.New()` constructor function, which takes said Options as arguments. Together, they allow code to look like this:

```go
// taken from `sendmail_test.go`, `TestTextMail`
var buf bytes.Buffer
sm := New(
	Subject("Cześć"),
	From("Michał", "me@example.com"),
	To("Ktoś", "info@example.com"),
	To("Ktoś2", "info2@example.com"),
	DebugOutput(&buf),
)
```

or like this:

```go
// taken from `options_test.go`, `TestChainingOptions`
m.SetSubject("Test subject").
	SetFrom("Dominik", "dominik@example.org").
	AppendTo("Dominik2", "dominik2@example.org").
	SetDebugOutput(&buf).
	SetSendmail("/bin/true")
```

Another subset of the commits moves the recently introduced global `Binary` variable, as well as the debug variable into the `Mail` type. As it turns out, global variables are a source of frustration in concurrent applications :-) You can see an example in the snippets above (`SetOutput`/`SetSendmail`).

I also started with HTML mails, but mixed-content (multipart, Text+HTML) isn't there yet. I'd like to harvest some code from [mailyak](https://github.com/domodwyer/mailyak), but the current license prevents me from that. Having a solid MIME multipart implementation would also allow attachments to the mail.

dmke committed:
Rework path to sendmail binary

It turns out, a global variable is not a great idea, so we move it
into the Mail object itself.

This is also introduces a new "chaninable" API: `m.SetSendmail()`
returns `m`, which allows chaining of further options (see following
commits).

dmke committed:
Update README [ci skip]

dmke committed:
Rework debug output

Don't depend on environment variable for debug output, instead let
the consumer decide where to print to.

dmke committed:
New chainable options: SetSubject, AppendTo, SetFrom

AppendTo might be a bit misleading, as it doesn't append the mail
to something, but appends the mail's To field. It is literally
a `To = append(To, ...)` construct.

dmke committed:
Additional Mail constructor

Accepts chainable Options as arguments.

dmke committed:
Remove debug delimiter

Makes testing unneccessarily hard.

dmke committed:
Allow either HTML or Text bodies

Place multipart mails (HTML+Text) on the roadmap.

dmke committed:
Travis: use gometalinter to perform linting

dmke committed:
Allow additional arguments to be passed to the sendmail program

meehow commented on sendmail.go:
```diff
@@ -87,19 +103,41 @@ func (m *Mail) Send() error {
 
 // WriteTo writes headers and content of the email to io.Writer
-func (m *Mail) WriteTo(wr io.Writer) error {
+func (m *Mail) WriteTo(wr io.Writer) error { // nolint: vet
```
If we are breaking compatibility with previous versions, then return values from this function can be also changed to these suggested by go vet.

dmke replied:
Yeah, I was about to ask about that, but forgot.

Since writing headers doesn't return the number of bytes written, the number we'd return could either be off by that amount, or always be 0. (We could also wrap the `wr` with something that counts the number of bytes written for us, but honestly, I don't think it matters that much...)

What do you think?

meehow commented on options.go:
```diff
@@ -0,0 +1,98 @@
+// AppendTo adds a recipient to the Mail. The name argument is the
+// "proper name" of the recipient and may be empty. The address must be
+// in the form "user@domain".
+func (m *Mail) AppendTo(name, address string) *Mail {
```
Here we can use `*mail.Address` directly, it's a build in type. I can imagine situation when somebody used `ParseAddress`, which returns `*mail.Address`, then he need to convert it to 2 strings and we convert it back to `*mail.Address`. That would be a bit over engineered. Using mail.Address as a function parameter doesn't bring that much complexity I think: `AppendTo(&mail.Address{Name: "someone", Address: "some@email"})`. @dmke what do you think?

dmke replied:
I'd be OK with that.

dmke committed:
Breaking: Satisfy govet and change Mail.WriteTo signature

WriteTo now returns (int64, error); however the returned number of bytes
written is always 0, since we don't get the actual number easily.

dmke committed:
Use *mail.Address as argument for {Append,}To and {Set,}From

dmke commented:
I've addressed your comments and added some commits.

I will now try to fix the issue of `m.Header.Write(wr)` not returning the number of bytes.

dmke committed:
Fix: Mail.WriteTo now returns the number of bytes written

dmke committed:
Tests for Mail.WriteTo

dmke commented:
How about that?

dmke commented:
FYI: I've worked on this on company time, so I need to move the code to my emplyers repository. This may require a new PR, depending on how Github handles moving repos :-)

dmke commented:
Any chance to get this merged in?

dmke committed:
AppendCC and AppendBCC

dmke committed:
Allow Append{To,CC,BCC} to receive multiple addresses

dmke commented:
Im closing this in favour of #3. This PR is (unfortunately) based on our `master` branch. For #3 I've moved the commits into a seperate branch.
